File: 7 Reverse Integer/C/solution.c
#include <limits.h>

#include "solution.h"

int reverse(int x) {
    int res = 0;

    while (x != 0) {
        int digit = x % 10;
        if (x > 0 && res > (INT_MAX - digit) / 10) {
            return 0;
        }
        if (x < 0 && res < (INT_MIN - digit) / 10) {
            return 0;
        }
        res = res * 10 + x % 10;
        x /= 10;
    }

    return res;
}

File: 7 Reverse Integer/Python/solution.py
from typing import Self

MIN_INT: int = -2**31
MAX_INT: int = 2**31 - 1

class Solution:
    def reverse(self: Self, x: int) -> int:
        absX: int = abs(x)
        multiplier: int = 1
        if x < 0:
            multiplier = -1
        
        res: int = 0
        while absX > 0:
            res = res * 10 + absX % 10
            absX //= 10

        if res < MIN_INT or res > MAX_INT:
            return 0

        return res * multiplier
    

if __name__ == "__main__":
    s: Solution = Solution()

    """_summary_
        Input: x = 123
        Output: 321
    """
    print(s.reverse(123))

    """_summary_
        Input: x = -123
        Output: -321
    """
    print(s.reverse(-123))

    """_summary_
        Input: x = 120
        Output: 21
    """
    print(s.reverse(120))